staging/src/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go

```diff
@@ -61,29 +53,124 @@ func NewLRUExpireCacheWithClock(maxSize int, clock Clock) *LRUExpireCache {
 type cacheEntry struct {
 	value      interface{}
 	expireTime time.Time
+
+	// If not nil, a computation is in progress. You must wait and then
+	// check finalValue.
+	wait              <-chan struct{}
+	finalValue        interface{}
+	finalValueOmitted bool
 }
 
 // Add adds the value to the cache at key with the specified maximum duration.
 func (c *LRUExpireCache) Add(key interface{}, value interface{}, ttl time.Duration) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	c.cache.Add(key, &cacheEntry{value, c.clock.Now().Add(ttl)})
+	et := c.clock.Now().Add(ttl)
+	c.cache.Add(key, &cacheEntry{
+		value:      value,
+		expireTime: et,
+	})
 }
 
-// Get returns the value at the specified key from the cache if it exists and is not
-// expired, or returns false.
+// Get returns the value at the specified key from the cache if it exists and
+// is not expired, or returns false. If the key is not in the cache because it
+// is still being computed, Get will return immediately as if the key were not
+// present. To wait, use GetOrWait.
 func (c *LRUExpireCache) Get(key interface{}) (interface{}, bool) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	e, ok := c.cache.Get(key)
-	if !ok {
+	ce, _ := c.getOrBegin(key, 0)
+	if ce == nil {
 		return nil, false
 	}
-	if c.clock.Now().After(e.(*cacheEntry).expireTime) {
-		c.cache.Remove(key)
+	if ce.wait != nil {
 		return nil, false
 	}
-	return e.(*cacheEntry).value, true
+	return ce.value, true
+}
+
+// ComputeFunc is used by GetOrWait to perform some work. You *must* exit if
+// the abort chan is closed/has a value. On failure, return a ttl of 0 to avoid
+// being cached; this will cause all waiting on the function to observe a cache miss.
+type ComputeFunc func(abort <-chan time.Time) (value interface{}, ttl time.Duration)
```

Comment:
It's not clear from the doc what time.Time value is sent on this channel. Why does this accept a time and not a context?